### PR TITLE
Members should be able to resend/revoke invitations they sent.

### LINKED ISSUE #8673: Members should be able to resend/revoke invitations they sent.

### PR DESCRIPTION
**Pull request by hakankaradis**
fixes #8673

### CODE REVIEW

File: client/app/lib/flux/teams/actions.coffee [hunk 1]
```diff
@@ -192,11 +192,13 @@ handlePendingInvitationUpdate = (account, action) ->
...
-    remote.api.JInvitation.some { '_id': account.get '_id' }, {}, (err, invitations) ->
-      invitation = invitations[0]
-      invitation.remove (err) ->
-        unless err
-          reactor.dispatch actions.DELETE_PENDING_INVITATION_SUCCESS, { account }
+    remote.api.JInvitation.revokeInvitation account, (err) ->
+
+      title = 'You are not authorized to revoke this invite.'
+
+      return new kd.NotificationView { title, 5000 }  if err
```
Comment (sinan) [inline client/app/lib/flux/teams/actions.coffee]: `{ 5000: 5000 }` wouldn't do what you plan, I guess :)

File: workers/social/lib/social/models/invitation.coffee [hunk 2]
```diff
@@ -222,6 +244,9 @@ module.exports = class JInvitation extends jraphical.Module
 
   createSingleInvite = (client, group, invitationData, end) ->
     { email, role, forceInvite, noEmail } = invitationData
+    { _id } = client.connection.delegate
+
+    invitationData.inviter_id = _id
```
Comment (sinan) [inline workers/social/lib/social/models/invitation.coffee]: `inviterId` camelCase please.

File: workers/social/lib/social/models/invitation.coffee [hunk 3]
```diff
@@ -98,6 +100,26 @@ module.exports = class JInvitation extends jraphical.Module
...
+        { roles } = permissionAndRoles
+
+        hasPermisson = 'admin' in roles or 'owner' in roles
+        { inviter_id, email, groupName } = invite
+
+        return callback 'noPermission'  if not hasPermisson and _id.toString() isnt inviter_id
+
+        JInvitation.remove { email, groupName }, (err) ->
+          return callback err
+
```
Comment (sinan) [inline workers/social/lib/social/models/invitation.coffee]: please add the backend test for this new method as well.

File: workers/social/lib/social/models/invitation.coffee [hunk 4]
```diff
@@ -98,6 +100,26 @@ module.exports = class JInvitation extends jraphical.Module
...
+      { connection: { delegate:account } } = client
+      account.fetchMyPermissions client, (err, permissionAndRoles) ->
+
+        { _id } = account
+        { roles } = permissionAndRoles
+
+        hasPermisson = 'admin' in roles or 'owner' in roles
+        { inviter_id, email, groupName } = invite
+
+        return callback 'noPermission'  if not hasPermisson and _id.toString() isnt inviter_id
```
Comment (sinan) [inline workers/social/lib/social/models/invitation.coffee]: we use `KodingError` wrapper for such errors in the backend please check similar places for an example. cc/ @cihangir

File: workers/social/lib/social/models/invitation.coffee (hunk 3 above)
Comment (hakankaradis) [inline workers/social/lib/social/models/invitation.coffee]: After I finalize this file regarding of style changes I will add.

File: client/app/lib/flux/teams/actions.coffee (hunk 1 above)
Comment (hakankaradis) [inline client/app/lib/flux/teams/actions.coffee]: oops...

File: workers/social/lib/social/models/invitation.coffee (hunk 3 above)
Comment (hakankaradis) [inline workers/social/lib/social/models/invitation.coffee]: 👍

**Comment by sinan** [comment]
@cihangir can you pls review the backend changes?

File: workers/social/lib/social/models/invitation.coffee [hunk 5]
```diff
@@ -25,7 +25,7 @@ module.exports = class JInvitation extends jraphical.Module
   @set
     permissions     :
       'send invitations'  : ['member']
-      'remove invitation' : ['admin']
+      'remove invitation' : ['member']
```
Comment (cihangir) [inline workers/social/lib/social/models/invitation.coffee]: changing a default permission requires maintenance in the db, i would suggest you to leave as is and implement a function like [this](https://github.com/koding/koding/blob/5956d6ca5f804e58452fe2aea548328fdb8acff8/workers/social/lib/social/models/group/index.coffee#L1150) and remove permit from function call signature, finally add that check into application itself. if user has permission to remove an invite do as is, if user does not have permission but regarding invitation has current user as it's inviter, again do that operation, else fail with an error message

File: workers/social/lib/social/models/invitation.coffee [hunk 6]
```diff
@@ -98,6 +100,29 @@ module.exports = class JInvitation extends jraphical.Module
...
         type        : String
         default     : -> 'member'
+      inviterId    :
+        type        : String
+
+  @revokeInvitation: permit 'remove invitation',
+    success: (client, invite, callback) ->
+
+      { connection: { delegate:account } } = client
+      account.fetchMyPermissions client, (err, permissionAndRoles) ->
```
Comment (cihangir) [inline workers/social/lib/social/models/invitation.coffee]: when you implement https://github.com/koding/koding/pull/8692/files#r73540359 you wont need this function

File: workers/social/lib/social/models/invitation.coffee [hunk 7]
```diff
@@ -98,6 +100,29 @@ module.exports = class JInvitation extends jraphical.Module
...
+
+        return callback new KodingError 'Couldn\'t fetch user permissions'  if err
+
+        { _id } = account
+        { roles } = permissionAndRoles
+
+        hasPermisson = 'admin' in roles or 'owner' in roles
+        { inviterId, email, groupName } = invite
```
Comment (cihangir) [inline workers/social/lib/social/models/invitation.coffee]: old invitations wont have `inviterId`handle that accordingly

File: workers/social/lib/social/models/invitation.coffee (hunk 5 above)
Comment (hakankaradis) [inline workers/social/lib/social/models/invitation.coffee]: @cihangir  As you said `remove permit from function call signature`
Are we able to call static instance that doesn't have permit from client side?
if not, how to make request for user who is member and the invitation is belong to her?

**Pull request merged**